core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java

```diff
@@ -486,66 +488,80 @@ public void addSplits(String tableName, SortedSet<Text> splits)
+          args.add(extent.endRow() == null ? EMPTY : TextUtil.getByteBuffer(extent.endRow()));
+          args.add(
+              extent.prevEndRow() == null ? EMPTY : TextUtil.getByteBuffer(extent.prevEndRow()));
+          splitsForTablet.getValue().forEach(split -> args.add(TextUtil.getByteBuffer(split)));
 
-        try {
-          handleFateOperation(() -> {
-            TFateInstanceType t = FateInstanceType.fromNamespaceOrTableName(tableName).toThrift();
-            TFateId opid = beginFateOperation(t);
-            executeFateOperation(opid, FateOperation.TABLE_SPLIT, args, Map.of(), false);
-            opids.add(new Pair<>(opid, splitsForTablet.getValue()));
-            return null;
-          }, tableName);
-        } catch (TableExistsException | NamespaceExistsException | NamespaceNotFoundException e) {
-          throw new RuntimeException(e);
-        }
-      }
-
-      // after all operations have been started, wait for them to complete
-      for (Pair<TFateId,List<Text>> entry : opids) {
-        final TFateId opid = entry.getFirst();
-        final List<Text> completedSplits = entry.getSecond();
+          try {
+            return handleFateOperation(() -> {
+              TFateInstanceType t = FateInstanceType.fromNamespaceOrTableName(tableName).toThrift();
+              TFateId opid = beginFateOperation(t);
+              executeFateOperation(opid, FateOperation.TABLE_SPLIT, args, Map.of(), false);
+              return new Pair<>(opid, splitsForTablet.getValue());
+            }, tableName);
+          } catch (TableExistsException | NamespaceExistsException | NamespaceNotFoundException
+              | AccumuloSecurityException | TableNotFoundException | AccumuloException e) {
+            throw new RuntimeException(e);
+          }
+          // wait for the fate operation to complete in a separate thread pool
+        }, startExecutor).thenApplyAsync(pair -> {
+          final TFateId opid = pair.getFirst();
+          final List<Text> completedSplits = pair.getSecond();
 
-        try {
-          String status = handleFateOperation(() -> waitForFateOperation(opid), tableName);
+          try {
+            String status = handleFateOperation(() -> waitForFateOperation(opid), tableName);
 
-          if (SPLIT_SUCCESS_MSG.equals(status)) {
-            completedSplits.forEach(splitsTodo::remove);
-          }
-        } catch (TableExistsException | NamespaceExistsException | NamespaceNotFoundException e) {
-          throw new RuntimeException(e);
-        } finally {
-          context.clearTableListCache();
-          // always finish table op, even when exception
-          if (opid != null) {
-            try {
-              finishFateOperation(opid);
-            } catch (Exception e) {
-              log.warn("Exception thrown while finishing fate table operation", e);
+            if (SPLIT_SUCCESS_MSG.equals(status)) {
+              synchronized (splitsTodo) {
```

Comment:
This is good to sync here for the background threads, but the access to `splitsTodo` is not synchronized when accessed in the foreground thread.  Since the foreground thread does not sync when calling `splitsTodo.isEmpty()` its not guaranteed it will see the changes made by background threads.  Could wrap `splitsTodo` using `Collections.synchronizedSortedSet()` when creating it.  Then the foreground thread would see changes made by the background threads.  If doing then then could also simplyfy the code here to a single line of `splitsTodo.removeAll(completedSplits);` and would not need the sync block.

Reply:
> Could wrap `splitsTodo` using `Collections.synchronizedSortedSet()` when creating it.

Good idea. I'll use the `synchronizedSortedSet` here.

> If doing that then could also simplyfy the code here to a single line of `splitsTodo.removeAll(completedSplits);` and would not need the sync block.

I had the same idea but my IDE gives the warning `Call to 'set.removeAll(list)' may work slowly` so thats kept the `foreach`.

Looked into it a bit and found that since we are removing items of a `List` from a `Set`, the `removeAll()` would contain multiple calls to `.contains()` which might make it less efficient. It might be more efficient to convert the `List` to a `Set` and then use `removeAll()` but I'm not sure. Alternatively, we could look into collecting straight to a `Set` in `mapSplitsToTablets()` instead of a `List`, then we could use `removeAll()` with both entities being a `Set`. I haven't looked into the implications of that though.

Reply:
> Looked into it a bit and found that since we are removing items of a List from a Set, the removeAll() would contain multiple calls to .contains() which might make it less efficient.

That is a good find.

> Alternatively, we could look into collecting straight to a Set in mapSplitsToTablets() instead of a List, then we could use removeAll() with both entities being a Set. I haven't looked into the implications of that though.

I don't think creating a set is worthwhile for this case.  I was thinking removeAll could avoid acquiring the lock over and over, but would not want to call contains on a list to avoid lock acquisition.  Looking at the impl in AbstractSet it seems to that when the list is smaller than the set.